lesson30/src/lib.rs

```diff
@@ -0,0 +1,585 @@
+     * Попытка снять больше чем есть на счете - ошибка.
+     */
+    Withdraw(NonZeroMoney), //снятие
+                            //перевол реализован через сумму операций Withdraw + Charge
+}
+
+pub trait OpsStorage {
+    fn transact(
+        &mut self,
+        ops: impl Iterator<Item = (AccountId, Operation)>,
+    ) -> Result<impl Iterator<Item = (OpId, AccountId, &Operation)>, Err>;
+
+    fn persist(&mut self, account_id: AccountId, op: Operation) -> Result<(OpId, &Operation), Err> {
+        self.transact(vec![(account_id, op)].into_iter())
+            .and_then(|mut iter| {
+                iter.next()
+                    .map(|(opt_id, _, op)| (opt_id, op))
+                    .ok_or("a transaction should return at least one operation".to_owned())
+            })
+    }
+    //should be O(N), where N - account ops
+    fn get_ops<'a, 'b>(
+        &'a self,
+        account_id: &'b AccountId,
+    ) -> Result<impl Iterator<Item = (OpId, &'a Operation)>, Err>;
+    //should be O(M), where M - all ops
+    fn get_history(&self) -> Result<impl Iterator<Item = (OpId, (AccountId, &Operation))>, Err>;
+}
+
+pub trait State {
+    fn transact<'a, 'b>(
+        &'a mut self,
+        ops: impl Iterator<Item = (AccountId, &'b Operation)>,
+    ) -> Result<impl Iterator<Item = (AccountId, &'a Account)>, Err>;
+
+    fn update<'a, 'b>(
+        &'a mut self,
+        account_id: AccountId,
+        op: &'b Operation,
+    ) -> Result<&'a Account, Err> {
+        self.transact(vec![(account_id, op)].into_iter())
+            .and_then(|mut iter| {
+                iter.next()
+                    .map(|(_, acc)| acc)
+                    .ok_or("a transaction should return at least one account".to_owned())
+            })
+    }
+    //should be O(N), where N - account ops
+    fn get_balance(&self, account_id: &AccountId) -> Result<&Account, Err>;
+}
+
+pub struct Bank<T: OpsStorage, S: State> {
+    storage: T,
+    state: S,
+}
+
+impl<T: OpsStorage, S: State> Bank<T, S> {
+    pub fn new(storage: T, state: S) -> Bank<T, S> {
+        Bank {
+            storage: storage,
+            state: state,
+        }
+    }
+
+    pub fn from<'a>(history: impl Iterator<Item = (OpId, (AccountId, &'a Operation))>) -> Bank<T, S>
+    where
+        S: Default,
+        T: Default,
+    {
+        let mut bank = Bank {
+            storage: T::default(),
+            state: S::default(),
+        };
+
+        for (_, (account_id, op)) in history {
+            let (_, op) = bank
+                .storage
+                .persist(account_id.clone(), op.clone())
+                .expect(format!("something wrong with history operation[{:?}]", op).as_str());
+            let ret = bank.state.update(account_id, op);
+
+            //let _ = ret.expect("History operation has an error");
+            ret.err()
+                .iter()
+                .for_each(|err| println!("History operation has an error[{:?}]", err));
```

Comment:
Пришлось прям подумать над этим местом, прежде чем понял, что там всегда будет только одна ошибка максимум. Предлагаю переписать вот так:

```suggestion
            if let Err(err) = ret {
                println!("History operation has an error[{err:?}]")
            }
```